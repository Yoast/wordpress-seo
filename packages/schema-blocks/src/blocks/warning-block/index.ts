import { WarningBlockAttributes } from "./warningBlockAttributes";

export { WarningBlockAttributes };
