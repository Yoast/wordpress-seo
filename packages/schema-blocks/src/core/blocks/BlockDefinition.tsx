import { ReactElement } from "react";
import { createElement, Fragment } from "@wordpress/element";
import {
	registerBlockType,
	BlockConfiguration,
	BlockEditProps,
	BlockSaveProps,
} from "@wordpress/blocks";
import { InspectorControls, BlockIcon } from "@wordpress/block-editor";
import { select } from "@wordpress/data";
import { PanelBody } from "@wordpress/components";

import BlockInstruction from "./BlockInstruction";
import Definition from "../Definition";
import BlockRootLeaf from "../../leaves/blocks/BlockRootLeaf";
import parse from "../../functions/blocks/parse";
import { registerBlockDefinition } from "./BlockDefinitionRepository";
import logger from "../../functions/logger";
import { openGeneralSidebar } from "../../functions/gutenberg/sidebar";

export interface RenderEditProps extends BlockEditProps<Record<string, unknown>> {
	clientId: string;
	name?: string;
}

export interface RenderSaveProps extends BlockSaveProps<Record<string, unknown>> {
	clientId?: string;
	className?: string;
}

export type MutableBlockConfiguration = {
	-readonly [K in keyof BlockConfiguration]: BlockConfiguration[K]
}

/**
 * BlockDefinition class.
 */
export default class BlockDefinition extends Definition {
	public static separatorCharacters = [ "b", "c", "d", "f", "g", "h", "k", "m", "z" ];
	public static parser = parse;

	public instructions: Record<string, BlockInstruction>;
	public tree: BlockRootLeaf;

	/**
	 * Renders editing the block.
	 *
	 * @param props The props.
	 *
	 * @returns The rendered block.
	 */
	edit( props: RenderEditProps ): ReactElement {
		// Force the sidebar open.
		if ( select( "core/block-editor" ).isBlockSelected( props.clientId ) ) {
			openGeneralSidebar( "edit-post/block", true );
		}

		const sidebarElements = this.sidebarElements( props );

		// Take the children directly to avoid creating too many Fragments.
		const elements = this.tree.children.map( ( leaf, i ) => leaf.edit( props, i ) ).filter( e => e !== null );

		if ( sidebarElements.length > 0 ) {
			const sidebarContainer =
				<InspectorControls key="sidebar-inspector-controls">
					<PanelBody>
						{ sidebarElements }
					</PanelBody>
				</InspectorControls>;

			elements.unshift( sidebarContainer );
		}

		if ( elements.length === 1 ) {
			return elements[ 0 ] as ReactElement;
		}

		return createElement( Fragment, { key: props.clientId }, elements.map( ( element, i ) => createElement( Fragment, { key: i }, element ) ) );
	}

	/**
	 * Renders the persisted block.
	 *
	 * @param props The props.
	 *
	 * @returns The rendered block.
	 */
	save( props: RenderSaveProps ): ReactElement {
		return this.tree.save( props );
	}

	/**
	 * Registers the block with Gutenberg.
	 */
	register(): void {
		const configuration = this.configuration() as MutableBlockConfiguration;
		const name = configuration.name as string;
		delete configuration.name;

		configuration.edit = props => this.edit( props );
		configuration.save = props => this.save( props );

		logger.debug( "registering block " + name );

		if ( configuration.icon && typeof configuration.icon === "string" && configuration.icon.startsWith( "<svg" ) ) {
			configuration.icon = this.createBlockIcon( configuration );
		}

		// Register the block to WordPress.
		registerBlockType( name, configuration );
		// Register the block with our own code.
		registerBlockDefinition( name, this );
	}

	/**
	 * Creates the sidebar elements.
	 *
	 * @param props The properties of the block to create a sidebar for.
	 *
	 * @returns The sidebar element to render.
	 */
	sidebarElements( props: RenderEditProps ): ReactElement[] {
		return Object.values( this.instructions )
			.map( ( instruction, index ) => <Fragment key={ instruction.id }>
				{ instruction.sidebar( props, index ) }
			</Fragment> )
			.filter( e => e !== null );
	}

	/**
	 * Creates a block icon.
	 *
	 * @param configuration The block configuration.
	 *
	 * @returns The sidebar element to render.
	 */
	private createBlockIcon( configuration: MutableBlockConfiguration ): ReactElement {
		const icon = <span
			className="yoast-schema-blocks-icon"
			dangerouslySetInnerHTML={ { __html: configuration.icon as string } }
		/>;

		return <BlockIcon icon={ icon } />;
	}
}
