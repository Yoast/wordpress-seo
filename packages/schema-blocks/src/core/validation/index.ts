import { BlockValidation } from "./BlockValidation";
import { BlockValidationResult, BlockPresence } from "./BlockValidationResult";
import { RequiredBlock } from "./RequiredBlock";
import { RecommendedBlock } from "./RecommendedBlock";
import { SuggestedBlockProperties } from "./SuggestedBlockProperties";

export {
	BlockPresence,
	BlockValidation,
	BlockValidationResult,
	RecommendedBlock,
	RequiredBlock,
	SuggestedBlockProperties,
};
