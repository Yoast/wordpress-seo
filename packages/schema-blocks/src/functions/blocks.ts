import { select } from "@wordpress/data";

/**
 * Gets attributes of inner blocks.
 *
 * @param clientId The client ID of the parent block.
 * @param blocks   A mapping of block name to attribute key.
 *
 * @returns An array contain block names and values.
 */
export function getInnerBlocksAttributes( clientId: string, blocks: Record<string, string> ): { name: string; value: unknown }[] {
	let innerBlocks = select( "core/block-editor" ).getBlock( clientId ).innerBlocks;
	innerBlocks     = innerBlocks.filter( block => block.name in blocks );

	const values = [];

	for ( const block of innerBlocks ) {
		const key   = blocks[ block.name ];
		values.push( { name: block.name, value: block.attributes[ key ] } );
	}

	return values;
}
