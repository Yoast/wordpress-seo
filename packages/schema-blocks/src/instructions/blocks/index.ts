import "./Block";
import "./ClassName";
import "./CurrencySelect";
import "./Date";
import "./Heading";
import "./InheritSidebar";
import "./InnerBlocks";
import "./RichText";
import "./Select";
import "./SidebarCheckbox";
import "./SidebarDuration";
import "./SidebarInput";
import "./SidebarSelect";
import "./TextInput";
import "./Title";
import "./Variation";
import "./VariationPicker";
import "./Duration";
