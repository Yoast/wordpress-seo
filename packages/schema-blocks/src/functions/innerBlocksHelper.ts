import { BlockInstance } from "@wordpress/blocks";
import { dispatch } from "@wordpress/data";
import recurseOverBlocks from "./blocks/recurseOverBlocks";

/**
 * Searches recursively in the inner blocks to get all instances of blocks whose name occurs in blockNames.
 *
 * @param blockInstance The array of blocks you're searching in.
 * @param blockNames    The names of the blocks you're searching for.
 *
 * @returns {BlockInstance[]} The block instances that have a name that occurs in blockNames.
 */
function getInnerblocksByName( blockInstance: BlockInstance, blockNames: string[] ): BlockInstance[] {
	return filterBlocksRecursively( blockInstance, block => blockNames.includes( block.name ) );
}

/**
 * Finds all innerblocks of a blockinstance that conform to the predicate.
 *
 * @param blockInstance The block whose innerblocks should be searched
 * @param predicate     The function to decide which blocks should be kept.
 * @returns {BlockInstance[]} The subset of innerblocks that conform to predicate.
 */
function filterBlocksRecursively( blockInstance: BlockInstance, predicate: ( blockInstance: BlockInstance ) => boolean ): BlockInstance[] {
	const foundBlocks: BlockInstance[] = [];

	recurseOverBlocks( blockInstance.innerBlocks, ( block: BlockInstance ) => {
		// Checks if the current block is one of the required types.
		if ( predicate( block ) ) {
			foundBlocks.push( block );
		}
	} );

	return foundBlocks;
}

/**
 * Maps the given callback function over all the blocks (including all innerBlocks) and returns the results as a flat array.
 *
 * @param blocks The blocks.
 * @param callback The callback function.
 *
 * @returns The transformed blocks, in a flat array.
 */
function mapBlocksRecursively<T>( blocks: BlockInstance[], callback: ( block: BlockInstance ) => T ): T[] {
	const result: T[] = [];
	recurseOverBlocks( blocks, ( block: BlockInstance ) => {
		// eslint-disable-next-line callback-return
		result.push( callback( block ) );
	} );
	return result;
}

/**
 * Gathers all the blocks, including inner blocks, in one array of blocks.
 *
 * @param blocks The list of blocks.
 *
 * @returns The list of all blocks.
 */
function getAllBlocks( blocks: BlockInstance[] ): BlockInstance[] {
	return mapBlocksRecursively( blocks, block => block );
}

/**
 * Inserts a block to the inner block.
 *
 * @param {BlockInstance} block    The block to insert.
 * @param {string}        clientId Id of the element to insert the block to.
 * @param {number}        index    The location of the block.
 */
function insertBlock( block: BlockInstance, clientId?: string, index?: number ): void {
	dispatch( "core/block-editor" ).insertBlock( block, index, clientId );
}

export { filterBlocksRecursively, getInnerblocksByName, mapBlocksRecursively, insertBlock, getAllBlocks };
