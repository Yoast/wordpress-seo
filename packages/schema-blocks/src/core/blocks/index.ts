import BlockDefinition from "./BlockDefinition";
import BlockInstruction from "./BlockInstruction";
import { BlockInstructionClass } from "./BlockInstructionClass";
import BlockLeaf from "./BlockLeaf";
import * as BlockDefinitionRepository from "./BlockDefinitionRepository";

export {
	BlockDefinition,
	BlockInstruction,
	BlockInstructionClass,
	BlockLeaf,
	BlockDefinitionRepository,
};
