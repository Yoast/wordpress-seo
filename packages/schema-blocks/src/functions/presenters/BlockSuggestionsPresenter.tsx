import { get } from "lodash";
import { YOAST_SCHEMA_BLOCKS_STORE_NAME } from "../redux";
import { ReactElement } from "react";
import { createBlock } from "@wordpress/blocks";
import { withSelect } from "@wordpress/data";
import { createElement } from "@wordpress/element";
import { insertBlock } from "../innerBlocksHelper";
import { isEmptyResult, isMissingResult, isValidResult } from "../validators/validateResults";
import { BlockValidationResult } from "../../core/validation";

type BlockSuggestionAddedProps = {
	blockTitle: string;
	isValid: boolean;
}

type BlockSuggestionProps = {
	suggestedBlockTitle: string;
	suggestedBlockName: string;
}

export type SuggestionDetails = BlockValidationResult & {
	title: string;
}

export interface BlockSuggestionsProps {
	heading: string;
	blockNames: string[];
}

export interface SuggestionsProps extends BlockSuggestionsProps {
	suggestions: SuggestionDetails[];
}

/**
 * Renders a block suggestion with the possibility to add one.
 *
 * @param blockTitle          The title to show.
 * @param suggestedBlockName  The name of the block to add.
 *
 * @returns The rendered block suggestion.
 */
function BlockSuggestion( { suggestedBlockTitle, suggestedBlockName }: BlockSuggestionProps ): ReactElement {
	/**
	 * Onclick handler for the remove block.
	 */
	const addBlockClick = () => {
		const blockToAdd = createBlock( suggestedBlockName );
		insertBlock( blockToAdd );
	};
	return (
		<li className="yoast-block-suggestion" key={ suggestedBlockTitle }>
			{ suggestedBlockTitle }
			{ /* eslint-disable-next-line react/jsx-no-bind */ }
			<button className="yoast-block-suggestion-button" onClick={ addBlockClick }> Add </button>
		</li>
	);
}

/**
 * Renders a block suggestion that has already been added.
 *
 * @param blockTitle The block title.
 * @param isValid    Is the added block valid.
 *
 * @returns The rendered element.
 */
function BlockSuggestionAdded( { blockTitle, isValid }: BlockSuggestionAddedProps ): ReactElement {
	const heroIconCheck = <svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="-2 3 18 16" stroke="currentColor" height="12" width="22">
		<path strokeLinecap="round" strokeLinejoin="round" strokeWidth={ 2.5 } d="M5 13l4 4L19 7" />
	</svg>;

	const checkmark = <span className="yoast-block-suggestion-checkmark">{ heroIconCheck }</span>;

	return (
		<li className="yoast-block-suggestion yoast-block-suggestion--added" key={ "BlockSuggestionAdded" + blockTitle }>
			{ blockTitle }
			{ isValid ? checkmark : null }
		</li>
	);
}

/**
 * Renders a list of block suggestions.
 * @param props The BlockValidationResults and the Blocks' titles.
 * @returns The appropriate Block Suggestion elements.
 */
export function PureBlockSuggestionsPresenter( { heading, suggestions, blockNames }: SuggestionsProps ): ReactElement {
	if ( ! suggestions || suggestions.length < 1 || ! blockNames || blockNames.length < 1 ) {
		return null;
	}

	return (
		<div>
			<div className="yoast-block-sidebar-title">{ heading }</div>
			<ul className="yoast-block-suggestions">
				{
					blockNames.map( ( blockName, index: number ) => {
						const suggestion = suggestions.find( sug => sug.name === blockName );

						const isValid = suggestion && isValidResult( suggestion.result );
						const isMissing = suggestion && isMissingResult( suggestion.result );
						const isEmpty = suggestion && isEmptyResult( suggestion.result );

						if ( isValid || isEmpty ) {
							// Show the validation result.
							return <BlockSuggestionAdded
								key={ index }
								isValid={ isValid }
								blockTitle={ suggestion.title }
							/>;
						}

						if ( isMissing ) {
							// Show the suggestion to add an instance of this block.
							return <BlockSuggestion
								key={ index }
								suggestedBlockTitle={ suggestion.title }
								suggestedBlockName={ suggestion.name }
							/>;
						}
					}, this )
				}
			</ul>
		</div>
	);
}

/**
 * Appends metadata to validation results retrieved from the store.
 *
 * @param props The props containing the blockNames we're interested in.
 *
 * @returns The props extended with suggestion data.
 */
export default withSelect<Partial<SuggestionsProps>, BlockSuggestionsProps, SuggestionsProps>( ( select, props: BlockSuggestionsProps ) => {
	const validations: BlockValidationResult[] =
		select( YOAST_SCHEMA_BLOCKS_STORE_NAME ).getValidationsForBlockNames( props.blockNames );

	const suggestionDetails = validations.map( validation => {
		const type = select( "core/blocks" ).getBlockType( validation.name );
		return {
			title: get( type, "title", "" ),
			...validation,
		};
	} );

	// The return object also includes properties from props.
	return {
		suggestions: suggestionDetails,
	};
} )( PureBlockSuggestionsPresenter );
