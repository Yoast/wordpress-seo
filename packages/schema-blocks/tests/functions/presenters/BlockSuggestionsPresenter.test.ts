import "../../matchMedia.mock";
import { mount } from "enzyme";
import * as renderer from "react-test-renderer";
import { createBlock } from "@wordpress/blocks";
import { BlockValidation, BlockValidationResult, BlockPresence } from "../../../src/core/validation";
import { PureBlockSuggestionsPresenter, SuggestionDetails, SuggestionsProps } from "../../../src/functions/presenters/BlockSuggestionsPresenter";
import { insertBlock } from "../../../src/functions/innerBlocksHelper";

jest.mock( "@wordpress/blocks", () => {
	return {
		createBlock: jest.fn(),
	};
} );

jest.mock( "../../../src/functions/validators", () => {
	return {
		getValidationResults: jest.fn( () => {
			return new BlockValidationResult( "1", "yoast/valid-block", -1, BlockPresence.Required, "Is not that present" );
		} ),
		getValidationResultForClientId: jest.fn( () => {
			return [
				new BlockValidationResult( "123", "yoast/added-to-content-valid", 1, BlockPresence.Required, "Is present" ),
			];
		} ),
	};
} );

jest.mock( "../../../src/functions/BlockHelper", () => {
	return {
		getBlockType: jest.fn( ( blockName: string )  => {
			if ( blockName === "yoast/nonexisting" ) {
				// eslint-disable-next-line no-undefined
				return undefined;
			}

			return {
				heading: "The required block",
			};
		} ),
		getHumanReadableBlockName: jest.fn( ( name: string ) => name ),
	};
} );

jest.mock( "../../../src/functions/innerBlocksHelper", () => {
	return {
		insertBlock: jest.fn(),
		getInnerblocksByName: jest.fn( ()  => {
			return [
				{
					name: "yoast/added-to-content",
					clientId: "existingBlockClientId",
				},
			];
		} ),
	};
} );

/**
 * Creates a mockery of a SuggestionDetails object
 * @param title      The Validated Block's title.
 * @param validation The validation result.
 * @returns Most of a SuggestionDetails object.
 */
function createSuggestion( title: string, validation: BlockValidationResult ): SuggestionDetails {
	const suggestion = ( validation as unknown as SuggestionDetails );
	suggestion.title = title;
	return suggestion;
}

describe( "The BlockSuggestionsPresenter class ", () => {
	it( "displays an [ Add ] link for missing required blocks", () => {
		// Arrange.
		const suggestion: SuggestionsProps = {
			heading: "Heading for Required Blocks",
			suggestions: [
				createSuggestion(
					"This is a missing block",
					new BlockValidationResult( null, "yoast/requiredBlock", BlockValidation.MissingRequiredBlock, BlockPresence.Required, null ),
				),
			],
			blockNames: [ "yoast/requiredBlock" ],
		};

		// Act.
		const tree = renderer
			.create( PureBlockSuggestionsPresenter( suggestion ) )
			.toJSON();

		// Assert.
		expect( tree ).toMatchSnapshot();
	} );

	it( "displays only the block title for blocks that aren't completed", () => {
		// Arrange.
		const suggestion: SuggestionsProps = {
			heading: "Heading for Required Blocks",
			suggestions: [
				createSuggestion(
					"This is an invalid required block without checkmark or add link",
					new BlockValidationResult( null, "yoast/requiredBlock", BlockValidation.MissingRequiredAttribute, BlockPresence.Required, null ),
				),
			],
			blockNames: [ "yoast/requiredBlock" ],
		};

		// Act.
		const tree = renderer
			.create( PureBlockSuggestionsPresenter( suggestion ) )
			.toJSON();

		// Assert.
		expect( tree ).toMatchSnapshot();
	} );

	it( "displays a checkmark for valid blocks", () => {
		// Arrange.
		const suggestion: SuggestionsProps = {
			heading: "Heading for Required Blocks",
			suggestions: [
				createSuggestion(
					"This is a valid required block with checkmark without add link",
					new BlockValidationResult( null, "yoast/requiredBlock", BlockValidation.Valid, BlockPresence.Required, null ),
				),
			],
			blockNames: [ "yoast/requiredBlock" ],
		};

		// Act.
		const tree = renderer
			.create( PureBlockSuggestionsPresenter( suggestion ) )
			.toJSON();

		// Assert.
		expect( tree ).toMatchSnapshot();
	} );

	it( "displays no suggestions if no suggestions are provided", () => {
		// Arrange.
		// eslint-disable-next-line max-len
		const suggestions: SuggestionsProps = {
			heading: "Recommended blocks",
			suggestions: [],
			blockNames: [ "yoast/recommendedBlock" ],
		};

		// Act.
		const tree = renderer
			.create( PureBlockSuggestionsPresenter( suggestions ) )
			.toJSON();

		// Assert.
		expect( tree ).toBeNull();
	} );

	it( "displays no suggestions if no blockNames are provided", () => {
		// Arrange.
		// eslint-disable-next-line max-len
		const validation = new BlockValidationResult( null, "yoast/recommendedBlock", BlockValidation.MissingRecommendedBlock, BlockPresence.Required, null );
		const suggestions: SuggestionDetails[] =
		[
			createSuggestion(
				"yoast/recommendedBlock",
				validation,
			),
		];

		// Act.
		const tree = renderer
			.create( PureBlockSuggestionsPresenter( { heading: "Recommended blocks", suggestions } as SuggestionsProps ) )
			.toJSON();

		// Assert.
		expect( tree ).toBeNull();
	} );

	it( "should add the block when the [ Add ] button is clicked.", () => {
		const suggestionsProps =  {
			heading: "Required blocks",
			suggestions: [
				createSuggestion( "yoast/not-added-to-content",
					BlockValidationResult.MissingBlock( "yoast/not-added-to-content", BlockPresence.Required ) ),
			],
			blockNames: [ "yoast/not-added-to-content" ],
		} as SuggestionsProps;

		const tree = mount( PureBlockSuggestionsPresenter( suggestionsProps ) );

		const addButton = tree.find( "button" ).first();

		addButton.simulate( "click" );

		expect( createBlock ).toHaveBeenCalled();
		expect( insertBlock ).toHaveBeenCalled();
	} );
} );
