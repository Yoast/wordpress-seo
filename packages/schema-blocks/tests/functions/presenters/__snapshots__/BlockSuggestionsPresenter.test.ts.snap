// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The BlockSuggestionsPresenter class  displays a checkmark for valid blocks 1`] = `
<div>
  <div
    className="yoast-block-sidebar-title"
  >
    Heading for Required Blocks
  </div>
  <ul
    className="yoast-block-suggestions"
  >
    <li
      className="yoast-block-suggestion yoast-block-suggestion--added"
    >
      This is a valid required block with checkmark without add link
      <span
        className="yoast-block-suggestion-checkmark"
      >
        <svg
          fill="none"
          height="12"
          stroke="currentColor"
          viewBox="-2 3 18 16"
          width="22"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 13l4 4L19 7"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={2.5}
          />
        </svg>
      </span>
    </li>
  </ul>
</div>
`;

exports[`The BlockSuggestionsPresenter class  displays an [ Add ] link for missing required blocks 1`] = `
<div>
  <div
    className="yoast-block-sidebar-title"
  >
    Heading for Required Blocks
  </div>
  <ul
    className="yoast-block-suggestions"
  >
    <li
      className="yoast-block-suggestion"
    >
      This is a missing block
      <button
        className="yoast-block-suggestion-button"
        onClick={[Function]}
      >
         Add 
      </button>
    </li>
  </ul>
</div>
`;

exports[`The BlockSuggestionsPresenter class  displays only the block title for blocks that aren't completed 1`] = `
<div>
  <div
    className="yoast-block-sidebar-title"
  >
    Heading for Required Blocks
  </div>
  <ul
    className="yoast-block-suggestions"
  >
    <li
      className="yoast-block-suggestion yoast-block-suggestion--added"
    >
      This is an invalid required block without checkmark or add link
    </li>
  </ul>
</div>
`;
